cluster/cluster.go

```diff
@@ -0,0 +1,437 @@
+package cluster
+
+import (
+	"context"
+	"io/ioutil"
+	"math/rand"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/hashicorp/memberlist"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/alertmanager/cluster/clusterpb"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Peer is a single peer in a gossip cluster.
+type Peer struct {
+	mlist    *memberlist.Memberlist
+	delegate *delegate
+
+	mtx    sync.RWMutex
+	states map[string]State
+	stopc  chan struct{}
+}
+
+const (
+	DefaultPushPullInterval = 20 * time.Second
+	DefaultGossipInterval   = 200 * time.Millisecond
+)
+
+func Join(
+	l log.Logger,
+	reg prometheus.Registerer,
+	bindAddr string,
+	advertiseAddr string,
+	knownPeers []string,
+	waitIfEmpty bool,
+	pushPullInterval time.Duration,
+	gossipInterval time.Duration,
+) (*Peer, error) {
+	bindHost, bindPortStr, err := net.SplitHostPort(bindAddr)
+	if err != nil {
+		return nil, err
+	}
+	bindPort, err := strconv.Atoi(bindPortStr)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid listen address")
+	}
+	var advertiseHost string
+	var advertisePort int
+
+	if advertiseAddr != "" {
+		var advertisePortStr string
+		advertiseHost, advertisePortStr, err = net.SplitHostPort(advertiseAddr)
+		if err != nil {
+			return nil, errors.Wrap(err, "invalid advertise address")
+		}
+		advertisePort, err = strconv.Atoi(advertisePortStr)
+		if err != nil {
+			return nil, errors.Wrap(err, "invalid advertise address, wrong port")
+		}
+	}
+
+	resolvedPeers, err := resolvePeers(context.Background(), knownPeers, advertiseAddr, net.Resolver{}, waitIfEmpty)
+	if err != nil {
+		return nil, errors.Wrap(err, "resolve peers")
+	}
+	level.Debug(l).Log("msg", "resolved peers to following addresses", "peers", strings.Join(resolvedPeers, ","))
+
+	// Initial validation of user-specified advertise address.
+	addr, err := calculateAdvertiseAddress(bindHost, advertiseHost)
+	if err != nil {
+		level.Warn(l).Log("err", "couldn't deduce an advertise address: "+err.Error())
+	} else if hasNonlocal(resolvedPeers) && isUnroutable(addr.String()) {
+		level.Warn(l).Log("err", "this node advertises itself on an unroutable address", "addr", addr.String())
+		level.Warn(l).Log("err", "this node will be unreachable in the cluster")
+		level.Warn(l).Log("err", "provide --cluster.advertise-address as a routable IP address or hostname")
+	}
+
+	// TODO(fabxc): generate human-readable but random names?
+	name, err := ulid.New(ulid.Now(), rand.New(rand.NewSource(time.Now().UnixNano())))
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Peer{
+		states: map[string]State{},
+		stopc:  make(chan struct{}),
+	}
+	p.delegate = newDelegate(l, reg, p)
+
+	cfg := memberlist.DefaultLANConfig()
```

Comment:
I don't remember entirely how to configure this, but memberlist exposes a bunch of metrics, but I think it has to be configured so it registers them with the Prometheus metrics registry. See: https://github.com/armon/go-metrics#sinks

Reply:
memberlist just writes to a global registry though :(

Reply:
After looking a bit more I'm against actually using their metrics. I doubt they are that useful for the average user and for the important ones we may be better of using our own ones.

The config has quite a lot of paramters which all scream that nothing about this is in line with Prometheus's semantics: https://github.com/armon/go-metrics/blob/master/start.go#L48
It may expose something well-formatted if configured right but it's probably going to be quite confusing semantically.

Reply:
Right I had a feeling like that, I stumbled upon the same findings last time I looked into this. Let's add our own metrics as applicable.

Reply:
One on "bytes_received/sent_total" seems useful though.